core/templates/dev/head/pages/exploration_editor/statistics_tab/LearnerActionRenderService.js

```diff
@@ -0,0 +1,158 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for rendering learner action HTML strings.
+ */
+
+oppia.factory('LearnerActionRenderService', [
+  'ACTION_TYPE_ANSWER_SUBMIT', 'ACTION_TYPE_EXPLORATION_START',
+  'ACTION_TYPE_EXPLORATION_QUIT', 'ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS',
+  function(
+      ACTION_TYPE_ANSWER_SUBMIT, ACTION_TYPE_EXPLORATION_START,
+      ACTION_TYPE_EXPLORATION_QUIT, ISSUE_TYPE_MULTIPLE_INCORRECT_SUBMISSIONS) {
+    var renderExplorationStartActionHTML = function(stateName) {
+      var htmlString =
+        'Started exploration at card "' + stateName + '".<br><br>';
+      return htmlString;
+    };
+
+    var renderExplorationQuitActionHTML = function(
+        stateName, timeSpentInStateSecs) {
+      var htmlString =
+        'Left the exploration after spending a total of ' +
+        timeSpentInStateSecs + ' seconds on card "' + stateName +
+        '".<br><br>';
+      return htmlString;
+    };
+
+    var renderContinueButtonSubmitActionHTML = function(
+        stateName, timeSpentInStateSecs) {
+      var htmlString =
+        'Pressed "Continue" to move to card "' + stateName + '" after ' +
+        timeSpentInStateSecs + ' seconds.<br><br>';
+      return htmlString;
+    };
+
+    var renderAnswerSubmitActionHTML = function(
+        answer, destStateName, timeSpentInStateSecs, currentStateName) {
+      var htmlString;
+      if (currentStateName === destStateName) {
+        htmlString =
+          'Submitted answer "' + answer + '" in card "' + currentStateName +
+          '".<br><br>';
+      } else {
+        htmlString =
+          'Submitted answer "' + answer + '" and moved to card "' +
+          destStateName + '" after spending ' + timeSpentInStateSecs +
+          ' seconds on card "' + currentStateName + '".<br><br>';
+      }
+      return htmlString;
+    };
+
+    var renderLearnerActionsTableForMultipleIncorrectIssue = function(
+        finalBlock) {
+      var index = finalBlock.length - 1;
+      var stateName =
+        finalBlock[index].actionCustomizationArgs.state_name.value;
+
+      var tableHTML = '<table><tr><th>Answer</th><th>Feedback</th></tr>';
+      for (var i = 0; i < index; i++) {
+        if (finalBlock[i].actionType !== ACTION_TYPE_ANSWER_SUBMIT) {
+          continue;
+        }
+        var answer = finalBlock[i].actionCustomizationArgs.answer.value;
+        var feedback = finalBlock[i].actionCustomizationArgs.feedback.value;
+        tableHTML +=
+          '<tr><td>' + answer + '</td><td>' + feedback + '</td></tr>';
+      }
+      tableHTML += '</table>';
+      return tableHTML;
+    };
+
+    var renderLearnerActionHTML = function(learnerAction) {
+      var actionType = learnerAction.actionType;
+      var custArgs = learnerAction.actionCustomizationArgs;
+      if (actionType === ACTION_TYPE_EXPLORATION_START) {
```

Comment:
This looks better suited for a mapping of some sort.

```javascript
var actionTypeRenderHtmlMap = {
  ACTION_TYPE_EXPLORATION_START: function(custArgs) {
    return renderExplorationStartActionHTML(custArgs.state_name.value);
  }, ACTION_TYPE_EXPLORATION_QUIT: function(custArgs) {
    return renderExplorationQuitActionHTML(
      custArgs.state_name.value, custArgs.time_spent_in_state_secs.value);
  }, // ...
};

actionTypeRenderHtmlMap[actionType](custArgs);
```

Reply:
I had some difficulties in making a mapping from a constant to a function, it didn't seem to work. I left it as is, but we can discuss it the next time.

Reply:
constants are treated as strings in JS, FYI -- I've run into this before. So, not sure there's a good alternative to if/else here.

Reply:
Acknowledged, that's an unfortunate limitation :(
Alright, I'm fine with this then!